### PR TITLE
Gen2 GH runner

### PR DESCRIPTION
**Pull request by zoete**
https://github.com/neondatabase/neon/issues/2158

### CODE REVIEW

**Comment by bayandin** [comment]
It looks like `Unsupported instrumentation profile format version` is caused by a mismatch in llvm versions — rust uses llvm-13 ([since 1.56](https://github.com/rust-lang/rust/pull/87570)), but we have llvm-10 in our image (should be fixed by https://github.com/neondatabase/build/pull/7)

**Comment by LizardWizzard** [comment]
Is it part of the https://github.com/neondatabase/neon/issues/2158? So I can pick it up, from there, right?

**Comment by bayandin** [comment]
> Is it part of the #2158? So I can pick it up, from there, right?

I work on this PR in the background (I need it for https://github.com/neondatabase/neon/pull/2071, so I do changes here from time to time, hopefully, it will be done today/tomorrow).

Sorry, it should have been assigned to me.


But I haven't checked/migrated any deploy-related jobs, so you can pick them @LizardWizzard

File: Dockerfile.compute-node
Comment (zoete) [inline Dockerfile.compute-node:1]: Will be used by follow-up merge: https://github.com/neondatabase/cloud/issues/1963 - currently inert.

File: .github/workflows/build_and_test.yml
Comment (bayandin) [inline .github/workflows/build_and_test.yml:13]: Is it removed for any particular reason?

File: .github/workflows/build_and_test.yml
Comment (bayandin) [inline .github/workflows/build_and_test.yml:22]: Let's use a specific tag, not just the latest one.
Also, we can move the image tag to the workflow env and use it `${{ env.RUST_TAG }}` to not to repeat it for each job

File: .github/workflows/build_and_test.yml
Comment (bayandin) [inline .github/workflows/build_and_test.yml:391]: Let's use a commit hash instead of a tag

File: .github/workflows/build_and_test.yml
Comment (bayandin) [inline .github/workflows/build_and_test.yml:409]: Do we use `@v1` for a particular reason? I think on other jobs we use `@v3`?

File: .github/workflows/build_and_test.yml
Comment (bayandin) [inline .github/workflows/build_and_test.yml:487]: Can we pin some versions/commits here as well?

File: .github/workflows/build_and_test.yml
Comment (bayandin) [inline .github/workflows/build_and_test.yml:587]: Can we use `rust` image here which comes with python?

File: Dockerfile.compute-node
Comment (bayandin) [inline Dockerfile.compute-node:27]: I'd say `stable-3.2` should be a build arg, and probably it should be a tag. 
And if we don't need to get it exactly from git, we can download https://download.osgeo.org/postgis/source/postgis-3.2.2.tar.gz, then `autogen.sh` won't be needed

File: .github/workflows/build_and_test.yml
Comment (bayandin) [inline .github/workflows/build_and_test.yml:399]: Do we really need to set `name` for each step?

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml:13]: Yes as kaniko & aws-cli don't have bash, I the run-shell for the jobs that need it

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml:409]: For v3 you need node, which didn't work for this image

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml:22]: Would conflict with lifecycle rule on ECR: Only keep 32 non latest images - latest will be behind an approval step on the build repository.

Comment (bayandin) [reply .github/workflows/build_and_test.yml:22]: > Would conflict with lifecycle rule on ECR: Only keep 32 non latest images

Wow, this is weird from them. Should we consider using other images storages? [GitHub container registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry), for example?

I'd expect running the same code to produce the same output (or about the same), which is nearly impossible to do with using `latest` tags. It also makes it harder to debug CI failures (when images switch occasionally)

File: .github/workflows/build_and_test.yml
Comment (bayandin) [reply .github/workflows/build_and_test.yml:409]: Ahhh, okay, worth adding a comment then.
Also we should use a tag for this image as well

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml:22]: No this is my added ECR rule, I guess I can put it to 128 non latest images but that would invalidate older pipelines, would that be preferred?

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml:481]: Both projects are strict on forward & backwards compatibility, similar to aws-cli I won't pin these.

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml:580]: To be done after this is merged as that would be a larger refactoring of the deployment script

File: .github/workflows/build_and_test.yml
Comment (bayandin) [reply .github/workflows/build_and_test.yml:22]: Is it 128 across in total or per image?

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml:407]: Kaniko is strict on forward & backwards compatibility, similar to aws-cli I won't pin this.

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml:22]: Per ECR repo, so per rust \ rustlegacy \ etc. Can go up much higher, but cost will increase.

File: .github/workflows/build_and_test.yml
Comment (bayandin) [reply .github/workflows/build_and_test.yml]: > Kaniko is strict on forward & backwards compatibility, similar to aws-cli I won't pin this.

Ok, can we pin a major version at least? Using just `debug` makes it unclear what is running (and why `debug`?)

Comment (zoete) [reply .github/workflows/build_and_test.yml]: Debug is needed else there is no shell to configure the ECR login, pinned on most recent release

File: .github/workflows/build_and_test.yml
Comment (bayandin) [reply .github/workflows/build_and_test.yml:481]: > Both projects are strict on forward & backwards compatibility, similar to aws-cli I won't pin these.

Does using the latest version of these tools benefit us somehow? I don't see advantages in it but can see potential problems.

Btw, the latest `golang` from the image can be incompatible with these tools.

File: .github/workflows/build_and_test.yml
Comment (bayandin) [reply .github/workflows/build_and_test.yml:22]: It depends on cost, but a stable build infrastructure is probably worth it

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml]: Updated with pins

File: .github/workflows/build_and_test.yml
Comment (zoete) [reply .github/workflows/build_and_test.yml:22]: What about this approach: https://github.com/neondatabase/build/pull/12 - then we have predictability but not inefficient levels of storage used

Comment (bayandin) [reply .github/workflows/build_and_test.yml:22]: Could you please explain what it does? 😄

Comment (zoete) [reply .github/workflows/build_and_test.yml:22]: Check if any images have tag "pinned" - if so, keep at least 3. - continue with rule #3. As posted in that pull, not sure what makes sense for workflow to "tag" an image as such, as it is infrequent.

Comment (zoete) [reply .github/workflows/build_and_test.yml]: Added the pinned tag, which for now is set manual. Will add this workflow to build readme

Comment (bayandin) [reply .github/workflows/build_and_test.yml]: Hmm, I think with `pinned` we'll have the same issues as with `latest` — workflow will pick up new images implicitly without any changes in the workflow itself.
I'm for having a new tag for each image version and letting the old one expire

Comment (zoete) [reply .github/workflows/build_and_test.yml]: Yes, however comparing the balancing act of keeping enough images (inefficient) that are almost all identical so we can have one specific numbered image available, say `rust:2746987948` - versus a workflow that will highlight what needs to be done before someone promotes an image to become the newest `rust:pinned` I would say the later one offers more control.

Comment (bayandin) [reply .github/workflows/build_and_test.yml]: Ok, let's switch it back to the `latest` and discuss it separately (to not block this PR further)

Comment (zoete) [reply .github/workflows/build_and_test.yml]: I got it on pinned now, which is outside of any automated workflow.

**Review by bayandin — APPROVED** [review]
If you have tested that the deploy part works as well, I'm 👍

File: .github/workflows/build_and_test.yml
Comment (bayandin) [reply .github/workflows/build_and_test.yml:427]: > Curious, how these access keys are now passed? I see that current builds have working cachepot, but don't see any env vars or build args like these anymore.

Access keys are "bundled" with runners we use (not sure I use the right terminology).

> We seem to miss GIT_VERSION since this PR, I'm trying to add it back.

Let me know if you need any help with it

Comment (SomeoneToIgnore) [reply .github/workflows/build_and_test.yml:427]: Thank you, I have been explained that IAM is set-up there, so that answers the first part.

And I've fixed the versions in https://github.com/neondatabase/neon/pull/2578 so we're good for now.

I was thinking to add some extra step after kaniko build to ensure that `pageserver --version` does not contain a default Docker version, but we need `docker run` for that.

**Pull request merged**
